atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraReadEteTest.java

```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesSeriallyIfDoesNotExceedLimit() {
+        IntStream.range(0, 50).forEach(i -> assertThat(readOneBatchOfSize(1)).hasSize(1));
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitFirstTime() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitSecondTimeWithSoftLimiting() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+        // The second read might actually be faster as the transaction/metadata is cached
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldNotBeAbleToReadLargeAmountsIfSoftLimitSleepWillBeMoreThanConfiguredBackoffTime() {
+        assertThatThrownBy(() -> readOneBatchOfSize(200))
+                .isInstanceOf(RateLimitExceededException.class)
+                .hasMessage("Rate limited. Available capacity has been exhausted.");
+    }
+
+    @Test
+    public void readRateLimitShouldBeRespectedByConcurrentReadingThreads() throws InterruptedException {
+        int numThreads = 5;
+        int numReadsPerThread = 10;
+        ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+        List<Future<List<Todo>>> futures = new ArrayList<>(numThreads);
+
+        long start = System.nanoTime();
+        IntStream.range(0, numThreads).forEach(i ->
+                futures.add(executorService.submit(() -> {
+                    List<Todo> results = new ArrayList<>(numReadsPerThread);
+                    IntStream.range(0, numReadsPerThread)
+                            .forEach(j -> results.addAll(readOneBatchOfSize(1)));
+                    return results;
+                })));
+        executorService.shutdown();
+        Preconditions.checkState(executorService.awaitTermination(30L, TimeUnit.SECONDS),
+                "Read tasks did not finish in 30s");
+        long timeTakenToReadInSeconds = TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - start);
+
+        assertThatAllReadsWereSuccessful(futures, numReadsPerThread);
+        long actualReadRate = (numThreads * numReadsPerThread * ONE_TODO_SIZE_IN_BYTES) / timeTakenToReadInSeconds;
+        assertThat(actualReadRate).isLessThan(readBytesPerSecond + (readBytesPerSecond / 10 /* to allow burst time */));
+    }
+
+    private void assertThatAllReadsWereSuccessful(List<Future<List<Todo>>> futures, int numReadsPerThread) {
+        AtomicInteger exceptionCounter = new AtomicInteger(0);
+        futures.forEach(future -> {
+            try {
+                assertThat(future.get()).hasSize(numReadsPerThread);
+            } catch (ExecutionException e) {
+                if (e.getCause() instanceof RateLimitExceededException) {
+                    exceptionCounter.getAndIncrement();
+                }
+            } catch (InterruptedException e) {
+                throw Throwables.propagate(e);
```

Comment:
nit: probably should just rethrow it or set the interrupted flag, but I guess for test code this is fine.

Reply:
Throwing a checked exception from the function makes it sad.